partiql-spi/src/main/kotlin/org/partiql/spi/value/ion/IonVariant.kt

```diff
@@ -54,127 +51,63 @@ internal class IonVariant(private var value: AnyElement) : Datum {
         return buffer.toByteArray()
     }
 
-    /**
-     * TODO!
-     */
     override fun lower(): Datum {
-        throw UnsupportedOperationException("VARIANT<ION> lower")
-    }
-
-    override fun getType(): PType = type
-
-    override fun isNull(): Boolean = value.isNull
-
-    override fun isMissing(): Boolean = false
-
-    override fun getString(): String = when (value.type) {
-        SYMBOL -> value.stringValue
-        STRING -> value.stringValue
-        else -> super.getString()
-    }
-
-    override fun getBoolean(): Boolean = when (value.type) {
-        BOOL -> value.booleanValue
-        else -> super.getBoolean()
-    }
-
-    // override fun getBytes(): ByteArray =  when (value.type) {
-    //     CLOB -> value.clobValue.copyOfBytes()
-    //     BLOB -> value.blobValue.copyOfBytes()
-    //     else -> super.getBytes()
-    // }
-    //
-    // override fun getByte(): Byte {
-    //     return super.getByte()
-    // }
-
-    override fun getDate(): Date {
-        return when (value.type) {
-            TIMESTAMP -> {
-                val ts = value.timestampValue
-                DateTimeValue.date(ts.year, ts.month, ts.day)
-            }
-            else -> super.getDate()
+        if (value.isNull) {
+            return Datum.nullValue(value.type.toPType())
         }
-    }
-
-    override fun getTime(): Time {
         return when (value.type) {
-            TIMESTAMP -> {
-                val ts = value.timestampValue
-                val tz = when (ts.localOffset) {
-                    null -> TimeZone.UnknownTimeZone
-                    else -> TimeZone.UtcOffset.of(ts.zHour, ts.zMinute)
+            STRING -> Datum.string(value.stringValue)
+            SYMBOL -> Datum.string(value.symbolValue)
+            BOOL -> Datum.bool(value.booleanValue)
+            CLOB -> Datum.clob(value.clobValue.copyOfBytes())
+            BLOB -> Datum.blob(value.blobValue.copyOfBytes())
+            TIMESTAMP -> Datum.timestamp(DateTimeValue.timestamp(value.timestampValue))
+            INT -> {
+                val bigInteger = value.bigIntegerValue
+                when {
+                    bigInteger < Int.MAX_VALUE.toBigInteger() && bigInteger > Int.MIN_VALUE.toBigInteger() -> Datum.integer(bigInteger.toInt())
+                    bigInteger < Long.MAX_VALUE.toBigInteger() && bigInteger > Long.MIN_VALUE.toBigInteger() -> Datum.bigint(bigInteger.toLong())
+                    else -> {
+                        val dec = bigInteger.toBigDecimal()
+                        Datum.decimal(dec, dec.precision(), 0)
+                    }
```

Comment:
nit: could use `IntElementSize` to simplify some of the logic here -- https://javadoc.io/doc/com.amazon.ion/ion-element/1.2.0/com/amazon/ionelement/api/IntElementSize.html. It only distinguishes between long and big_integer (not sure if that's intentional) but it could help eliminate some of the `.toBigInteger` conversions.